### PR TITLE
SparkFun Qwiic Pro Micro USB-C Atmega32u4

### LINKED ISSUE #52: Add support for: SparkFun Qwiic Pro Micro - USB-C (ATmega32U4)
Shop page: https://www.sparkfun.com/products/15795
Datasheet: https://cdn.sparkfun.com/datasheets/Dev/Arduino/Boards/ATMega32U4.pdf
Addons: https://cdn.sparkfun.com/assets/learn_tutorials/1/2/1/SparkFunBoard_Addons_Jan_16.zip

As discussed it would be nice to update avr-hal with this board. I tested entirely using the leonardo crate and I couldn't find any difference between the Leonardo and this board.

Here is the pin file provided in the addons for this board (included in the zip file above):
```
/*
  pins_arduino.h - Pin definition functions for Arduino
  Part of Arduino - http://www.arduino.cc/

  Copyright (c) 2007 David A. Mellis

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

  $Id: wiring.h 249 2007-02-03 16:52:51Z mellis $
*/

#ifndef Pins_Arduino_h
#define Pins_Arduino_h

#include <avr/pgmspace.h>

// Workaround for wrong definitions in "iom32u4.h".
// This should be fixed in the AVR toolchain.
#undef UHCON
#undef UHINT
#undef UHIEN
#undef UHADDR
#undef UHFNUM
#undef UHFNUML
#undef UHFNUMH
#undef UHFLEN
#undef UPINRQX
#undef UPINTX
#undef UPNUM
#undef UPRST
#undef UPCONX
#undef UPCFG0X
#undef UPCFG1X
#undef UPSTAX
#undef UPCFG2X
#undef UPIENX
#undef UPDATX
#undef TCCR2A
#undef WGM20
#undef WGM21
#undef COM2B0
#undef COM2B1
#undef COM2A0
#undef COM2A1
#undef TCCR2B
#undef CS20
#undef CS21
#undef CS22
#undef WGM22
#undef FOC2B
#undef FOC2A
#undef TCNT2
#undef TCNT2_0
#undef TCNT2_1
#undef TCNT2_2
#undef TCNT2_3
#undef TCNT2_4
#undef TCNT2_5
#undef TCNT2_6
#undef TCNT2_7
#undef OCR2A
#undef OCR2_0
#undef OCR2_1
#undef OCR2_2
#undef OCR2_3
#undef OCR2_4
#undef OCR2_5
#undef OCR2_6
#undef OCR2_7
#undef OCR2B
#undef OCR2_0
#undef OCR2_1
#undef OCR2_2
#undef OCR2_3
#undef OCR2_4
#undef OCR2_5
#undef OCR2_6
#undef OCR2_7

#define NUM_DIGITAL_PINS  31
#define NUM_ANALOG_INPUTS 12

#define TX_RX_LED_INIT  DDRD |= (1<<5), DDRB |= (1<<0)
#define TXLED0          PORTD |= (1<<5)
#define TXLED1          PORTD &= ~(1<<5)
#define RXLED0          PORTB |= (1<<0)
#define RXLED1          PORTB &= ~(1<<0)

static const uint8_t SDA = 2;
static const uint8_t SCL = 3;
#define LED_BUILTIN 13
#define LED_BUILTIN_RX 17					
#define LED_BUILTIN_TX 30

// Map SPI port to 'new' pins D14..D17
static const uint8_t SS   = 17;
static const uint8_t MOSI = 16;
static const uint8_t MISO = 14;
static const uint8_t SCK  = 15;

// Mapping of analog pins as digital I/O
// A6-A11 share with digital pins
static const uint8_t A0 = 18;
static const uint8_t A1 = 19;
static const uint8_t A2 = 20;
static const uint8_t A3 = 21;
static const uint8_t A4 = 22;
static const uint8_t A5 = 23;
static const uint8_t A6 = 24;   // D4
static const uint8_t A7 = 25;   // D6
static const uint8_t A8 = 26;   // D8
static const uint8_t A9 = 27;   // D9
static const uint8_t A10 = 28;  // D10
static const uint8_t A11 = 29;  // D12

#define digitalPinToPCICR(p)    ((((p) >= 8 && (p) <= 11) || ((p) >= 14 && (p) <= 17) || ((p) >= A8 && (p) <= A10)) ? (&PCICR) : ((uint8_t *)0))
#define digitalPinToPCICRbit(p) 0
#define digitalPinToPCMSK(p)    ((((p) >= 8 && (p) <= 11) || ((p) >= 14 && (p) <= 17) || ((p) >= A8 && (p) <= A10)) ? (&PCMSK0) : ((uint8_t *)0))
#define digitalPinToPCMSKbit(p) ( ((p) >= 8 && (p) <= 11) ? (p) - 4 : ((p) == 14 ? 3 : ((p) == 15 ? 1 : ((p) == 16 ? 2 : ((p) == 17 ? 0 : (p - A8 + 4))))))

//  __AVR_ATmega32U4__ has an unusual mapping of pins to channels
extern const uint8_t PROGMEM analog_pin_to_channel_PGM[];
#define analogPinToChannel(P)  ( pgm_read_byte( analog_pin_to_channel_PGM + (P) ) )

#define digitalPinToInterrupt(p) ((p) == 0 ? 2 : ((p) == 1 ? 3 : ((p) == 2 ? 1 : ((p) == 3 ? 0 : ((p) == 7 ? 4 : NOT_AN_INTERRUPT)))))

#ifdef ARDUINO_MAIN

// On the Arduino board, digital pins are also used
// for the analog output (software PWM).  Analog input
// pins are a separate set.

// ATMEL ATMEGA32U4 / ARDUINO LEONARDO
//
// D0               PD2                 RXD1/INT2
// D1               PD3                 TXD1/INT3
// D2               PD1     SDA         SDA/INT1
// D3#              PD0     PWM8/SCL    OC0B/SCL/INT0
// D4       A6      PD4                 ADC8
// D5#              PC6     ???         OC3A/#OC4A
// D6#      A7      PD7     FastPWM     #OC4D/ADC10
// D7               PE6                 INT6/AIN0
//
// D8       A8      PB4                 ADC11/PCINT4
// D9#      A9      PB5     PWM16       OC1A/#OC4B/ADC12/PCINT5
// D10#     A10     PB6     PWM16       OC1B/0c4B/ADC13/PCINT6
// D11#             PB7     PWM8/16     0C0A/OC1C/#RTS/PCINT7
// D12      A11     PD6                 T1/#OC4D/ADC9
// D13#             PC7     PWM10       CLK0/OC4A
//
// A0       D18     PF7                 ADC7
// A1       D19     PF6                 ADC6
// A2       D20     PF5                 ADC5
// A3       D21     PF4                 ADC4
// A4       D22     PF1                 ADC1
// A5       D23     PF0                 ADC0
//
// New pins D14..D17 to map SPI port to digital pins
//
// MISO     D14     PB3                 MISO,PCINT3
// SCK      D15     PB1                 SCK,PCINT1
// MOSI     D16     PB2                 MOSI,PCINT2
// SS       D17     PB0                 RXLED,SS/PCINT0
//
// Connected LEDs on board for TX and RX
// TXLED    D30     PD5                 XCK1
// RXLED    D17     PB0
// HWB              PE2                 HWB

// these arrays map port names (e.g. port B) to the
// appropriate addresses for various functions (e.g. reading
// and writing)
const uint16_t PROGMEM port_to_mode_PGM[] = {
    NOT_A_PORT,
    NOT_A_PORT,
    (uint16_t) &DDRB,
    (uint16_t) &DDRC,
    (uint16_t) &DDRD,
    (uint16_t) &DDRE,
    (uint16_t) &DDRF,
};

const uint16_t PROGMEM port_to_output_PGM[] = {
    NOT_A_PORT,
    NOT_A_PORT,
    (uint16_t) &PORTB,
    (uint16_t) &PORTC,
    (uint16_t) &PORTD,
    (uint16_t) &PORTE,
    (uint16_t) &PORTF,
};

const uint16_t PROGMEM port_to_input_PGM[] = {
    NOT_A_PORT,
    NOT_A_PORT,
    (uint16_t) &PINB,
    (uint16_t) &PINC,
    (uint16_t) &PIND,
    (uint16_t) &PINE,
    (uint16_t) &PINF,
};

const uint8_t PROGMEM digital_pin_to_port_PGM[] = {
    PD, // D0 - PD2
    PD, // D1 - PD3
    PD, // D2 - PD1
    PD, // D3 - PD0
    PD, // D4 - PD4
    PC, // D5 - PC6
    PD, // D6 - PD7
    PE, // D7 - PE6
    
    PB, // D8 - PB4
    PB, // D9 - PB5
    PB, // D10 - PB6
    PB, // D11 - PB7
    PD, // D12 - PD6
    PC, // D13 - PC7
    
    PB, // D14 - MISO - PB3
    PB, // D15 - SCK - PB1
    PB, // D16 - MOSI - PB2
    PB, // D17 - SS - PB0
    
    PF, // D18 - A0 - PF7
    PF, // D19 - A1 - PF6
    PF, // D20 - A2 - PF5
    PF, // D21 - A3 - PF4
    PF, // D22 - A4 - PF1
    PF, // D23 - A5 - PF0
    
    PD, // D24 / PD5
    PD, // D25 / D6 - A7 - PD7
    PB, // D26 / D8 - A8 - PB4
    PB, // D27 / D9 - A9 - PB5
    PB, // D28 / D10 - A10 - PB6
    PD, // D29 / D12 - A11 - PD6
    PD, // D30 / TX Led - PD5
};

const uint8_t PROGMEM digital_pin_to_bit_mask_PGM[] = {
    _BV(2), // D0 - PD2
    _BV(3), // D1 - PD3
    _BV(1), // D2 - PD1
    _BV(0), // D3 - PD0
    _BV(4), // D4 - PD4
    _BV(6), // D5 - PC6
    _BV(7), // D6 - PD7
    _BV(6), // D7 - PE6
    
    _BV(4), // D8 - PB4
    _BV(5), // D9 - PB5
    _BV(6), // D10 - PB6
    _BV(7), // D11 - PB7
    _BV(6), // D12 - PD6
    _BV(7), // D13 - PC7
    
    _BV(3), // D14 - MISO - PB3
    _BV(1), // D15 - SCK - PB1
    _BV(2), // D16 - MOSI - PB2
    _BV(0), // D17 - SS - PB0
    
    _BV(7), // D18 - A0 - PF7
    _BV(6), // D19 - A1 - PF6
    _BV(5), // D20 - A2 - PF5
    _BV(4), // D21 - A3 - PF4
    _BV(1), // D22 - A4 - PF1
    _BV(0), // D23 - A5 - PF0
    
	  _BV(4), // D24 / D4 - A6 - PD4
    _BV(7), // D25 / D6 - A7 - PD7
    _BV(4), // D26 / D8 - A8 - PB4
    _BV(5), // D27 / D9 - A9 - PB5
    _BV(6), // D28 / D10 - A10 - PB6
    _BV(6), // D29 / D12 - A11 - PD6
    _BV(5), // D30 / TX Led - PD5
};

const uint8_t PROGMEM digital_pin_to_timer_PGM[] = {
    NOT_ON_TIMER,   
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    TIMER0B,        /* 3 */
    NOT_ON_TIMER,
    TIMER3A,        /* 5 */
    TIMER4D,        /* 6 */
    NOT_ON_TIMER,   
    
    NOT_ON_TIMER,   
    TIMER1A,        /* 9 */
    TIMER1B,        /* 10 */
    TIMER0A,        /* 11 */
    
    NOT_ON_TIMER,   
    TIMER4A,        /* 13 */
    
    NOT_ON_TIMER,   
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,

    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
    NOT_ON_TIMER,
};

const uint8_t PROGMEM analog_pin_to_channel_PGM[] = {
    7,  // A0               PF7                 ADC7
    6,  // A1               PF6                 ADC6    
    5,  // A2               PF5                 ADC5    
    4,  // A3               PF4                 ADC4
    1,  // A4               PF1                 ADC1    
    0,  // A5               PF0                 ADC0    
    8,  // A6       D4      PD4                 ADC8
    10, // A7       D6      PD7                 ADC10
    11, // A8       D8      PB4                 ADC11
    12, // A9       D9      PB5                 ADC12
    13, // A10      D10     PB6                 ADC13
    9   // A11      D12     PD6                 ADC9
};

#endif /* ARDUINO_MAIN */

// These serial port names are intended to allow libraries and architecture-neutral
// sketches to automatically default to the correct port name for a particular type
// of use.  For example, a GPS module would normally connect to SERIAL_PORT_HARDWARE_OPEN,
// the first hardware serial port whose RX/TX pins are not dedicated to another use.
//
// SERIAL_PORT_MONITOR        Port which normally prints to the Arduino Serial Monitor
//
// SERIAL_PORT_USBVIRTUAL     Port which is USB virtual serial
//
// SERIAL_PORT_LINUXBRIDGE    Port which connects to a Linux system via Bridge library
//
// SERIAL_PORT_HARDWARE       Hardware serial port, physical RX & TX pins.
//
// SERIAL_PORT_HARDWARE_OPEN  Hardware serial ports which are open for use.  Their RX & TX
//                            pins are NOT connected to anything by default.
#define SERIAL_PORT_MONITOR        Serial
#define SERIAL_PORT_USBVIRTUAL     Serial
#define SERIAL_PORT_HARDWARE       Serial1
#define SERIAL_PORT_HARDWARE_OPEN  Serial1

// Alias SerialUSB to Serial
#define SerialUSB SERIAL_PORT_USBVIRTUAL

#endif /* Pins_Arduino_h */
```

### PR DESCRIPTION
**Pull request by cecton**
Hello @Rahix 

I just started making the sparkfun-pro-micro crate by copying and renaming everything from the arduino-leonardo one (26698a0a191).

> @cecton, for the pins, just reference the [schematic for your board](https://cdn.sparkfun.com/assets/4/4/f/2/a/Qwiic_Pro_Micro_V2_0_USB_C_Schematic.pdf). It shows quite well which MCU pin is named what on the PCB.

It honestly took me some time to see it on the schematics :sweat_smile: I used the right most diagram.

It doesn't really show D11 to D13 but I supposed this is assumed because of the order? So I did the same with A4-A5 (Please double check that b66921e87)

Closes #52

### CODE REVIEW

File: boards/sparkfun-pro-micro/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "sparkfun-pro-micro"
+version = "0.1.0"
+authors = ["Rahix <rahix@rahix.de>"]
```
Comment (Rahix) [inline boards/sparkfun-pro-micro/Cargo.toml:4]: That should be you :)

**Review by Rahix — COMMENTED** [review]
Looks good, I have two more things:

1. Please add a section in README, like there is for all the other boards.
2. Please add the board to `.travis.yml` so CI test-builds all examples.

File: boards/sparkfun-pro-micro/Cargo.toml (hunk 1 above)
Comment (cecton) [reply boards/sparkfun-pro-micro/Cargo.toml]: Yeay \o/ Thanks!!

**Comment by cecton** [comment]
Ok I have changed what you asked but there are 2 things I would like to discuss with you:

1. I noticed in the examples of the leonardo that the port F is actually used in the example leonardo-adc. But then why it isn't in the pins file instead? Because if it were, the leonardo and the promicro would be the exact same board.
2. The example i2cdetect actually crashes on the pro micro. I'm not certain and I'm not sure how I can debug it. I think it crashes because when I run it with the screen connected on qwiic, I get this:

```
Write direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

Read direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c
```

While when the screen is not connected I get this:

```
Write direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

Read direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --
```

I also tried to light a led right after the read and it doesn't work. So I guess it's crashing somehow.

**Comment by cecton** [comment]
I couldn't reproduce the error on the CI so I did an rustup update and now nothing compiles XD

**Comment by Rahix** [comment]
@cecton, you'll need to rebase your branch on latest master, then you should be able to reproduce the error.  From the CI log, I think you're missing the prescaler value for PWM.  Should be easy to fix, look at the current version of the example in the leonardo crate.

**Comment by cecton** [comment]
I thought I did update but I updated on my master ...

**Comment by cecton** [comment]
It's ready

**Review by Rahix — APPROVED** [review]

**Comment by cecton** [comment]
@Rahix hey o/ if you have some time, could you answer my questions? I think there are improvements to do, I can open separate tickets since 2. is probably a bug but I'm not sure about 1.

**Comment by Rahix** [comment]
> I noticed in the examples of the leonardo that the port F is actually used in the example leonardo-adc. But then why it isn't in the pins file instead?

Right, the real reason is that I didn't want to add it until ADC was implemented ... Which it now is, so this should definitely be fixed!

> Because if it were, the leonardo and the promicro would be the exact same board.

Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in [issue #64](https://github.com/Rahix/avr-hal/issues/64#issuecomment-683676111) and implemented in [pr #69](https://github.com/Rahix/avr-hal/pull/69))? I.e. add a feature-flag for the promicro to `arduino-leonardo` and gate any promicro specific functionality behind that (if there is any).

> The example i2cdetect actually crashes on the pro micro. I'm not certain and I'm not sure how I can debug it. I think it crashes because when I run it with the screen connected on qwiic, [...]

This is most likely not a bug in the i2cdetect code but caused by an unfortunate design flaw in I2C.  From the Linux `i2cdetect` man-page:

```text
WARNING
        This program can confuse your I2C bus, cause data loss and worse! 
```

And an [excerpt from the source](https://git.kernel.org/pub/scm/utils/i2c-tools/i2c-tools.git/tree/tools/i2cdetect.c#n100):

```c
/* Probe this address */
switch (cmd) {
default: /* MODE_QUICK */
  /* This is known to corrupt the Atmel AT24RF08
     EEPROM */
  res = i2c_smbus_write_quick(file,
        I2C_SMBUS_WRITE);
  break;
case MODE_READ:
  /* This is known to lock SMBus on various
     write-only chips (mainly clock chips) */
  res = i2c_smbus_read_byte(file);
  break;
}
```

Essentially, the problem is that I2C does not have a dedicated operation for probing addresses.  Instead, `i2cdetect` implementations attempt to read from each address / write to each address and whenever a slave device responds with `ACK`, it knows something is connected at that address.  However, some devices can't deal with random reads (e.g. some write-only devices) or writes (e.g. because the device interprets the written value) and enter a weird state where they lock up the bus.  Most likely, that's why it's hanging for you ...

**Comment by cecton** [comment]
> Right, the real reason is that I didn't want to add it until ADC was implemented ... Which it now is, so this should definitely be fixed!

Ok I will do it! :)

> Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in issue #64 and implemented in pr #69)? I.e. add a feature-flag for the promicro to arduino-leonardo and gate any promicro specific functionality behind that (if there is any). 

I guess I can try to do that too. The code is exactly the same for leonardo and this promicro.

> This is most likely not a bug in the i2cdetect code but caused by an unfortunate design flaw in I2C. From the Linux `i2cdetect` man-page:

Linux? Sorry I meant the i2cdetect function of avr-hal: https://rahix.github.io/avr-hal/atmega32u4_hal/i2c/struct.I2c.html#method.i2cdetect

**Comment by Rahix** [comment]
> Linux? Sorry I meant the i2cdetect function of avr-hal

Sure, I just referred to Linux as they do the exact same thing and it's better documented over there ;)

**Comment by cecton** [comment]
> > Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in issue #64 and implemented in pr #69)? I.e. add a feature-flag for the promicro to arduino-leonardo and gate any promicro specific functionality behind that (if there is any).
> 
> I guess I can try to do that too. The code is exactly the same for leonardo and this promicro.

I checked the code and I see the point. I said that the leonardo and promicro are exactly the same but I lied. The promicro has only 2 leds. Does that mean I get the leonaro a feature flag of the pro micro?

I also realized that I might have added too many pins. i.e. there are no pin A4 and A5 that I can see on the board! I probably made a mistake.

```rust
        /// `A0`
        pub a0: portf::pf7::PF7,
        /// `A1`
        pub a1: portf::pf6::PF6,
        /// `A2`
        pub a2: portf::pf5::PF5,
        /// `A3`
        pub a3: portf::pf4::PF4,
        /// `A4`
        pub a4: portf::pf1::PF1,
        /// `A5`
        pub a5: portf::pf0::PF0,
```

(Also not existing: D11-D13 but I added them in the code)

All of this makes me realize that promicro is really a subset of leonardo. If I make promicro a feature flag I will have put only `#[cfg(not(feature = "promicro"))]` and if I make leonardo a feature flag I will have to put only `#[cfg(feature = "leonardo")]` which is simpler (especially if we add more boards. But on the other hand, leonardo was first but, again, everything is in `0.` version so you can afford to do it easily.

Or maybe we can keep all the duplicated code for now. Using the build.rs it is probably possible to copy the code on fly during the compilation and have separate crates.

**Comment by Rahix** [comment]
I would argue for the `#[cfg(not(feature = "promicro"))]` method.  This way the crate by default is everything leonardo and each feature flag just denotes the deviations.  Of course, for more complex scenarios one would need to carefully consider which board to use as the "base" but here I think the Leonardo is a safe bet.

**Comment by Rahix** [comment]
> Or maybe we can keep all the duplicated code for now. Using the build.rs it is probably possible to copy the code on fly during the compilation and have separate crates.

Well worst case we shouldn't use build.rs but go with `include!()` but I would very much like to avoid going that route ... If I understand you correctly, the two boards are similar enough that it makes sense to put them in a single crate, similar to _Arduino Uno_ and _Arduino Nano_.

**Comment by cecton** [comment]
I think they're similar only because the base --here the microchip-- is the same. If I understand things correctly, some things are connected, some are not, but the base should probably be the microschip and not a board.

I think the feature was okay for arduino uno/nano but it feels wrong to "hide" sparkfun's product behind a crate called "arduino". This only increases the confusion many people (including me) have about what is exactly an Arduino. For that reason I do prefer promicro getting its own crate.

**Pull request merged**